pkg/util/fxutil/provide_comp.go

```diff
@@ -0,0 +1,299 @@
+// into the fx constructor graph.
+//
+// For example, given:
+//
+//	type Provides struct {
+//	    My MyComponent
+//	}
+//	type Requires struct {
+//	    Dep MyDependency
+//	}
+//	func NewComponent(reqs Requires) Provides { ... }
+//
+// then:
+//
+//	ProvideComponentConstructor(NewComponent)
+//
+// will create these anonymous types:
+//
+//	type FxAwareProvides struct {
+//	    fx.Out
+//	    My MyComponent
+//	}
+//	type FxAwareRequires struct {
+//	    fx.In
+//	    Dep MyDependency
+//	}
+//
+// and essentially make this happen:
+//
+//	fx.Provide(func(fxr FxAwareRequires) fxp FxAwareProvides {
+//	    r := makePlainReqs(fxr)
+//	    p := NewComponent(r)
+//	    return makeFxAwareProvides(p)
+//	})
+
+var (
+	compOutType = reflect.TypeOf((*compdef.Out)(nil)).Elem()
+	fxInType    = reflect.TypeOf(fx.In{})
+	fxOutType   = reflect.TypeOf(fx.Out{})
+)
+
+// ProvideComponentConstructor takes a Component constructor and makes it fx-aware, then fx.Provide's it
+func ProvideComponentConstructor(compCtorFunc interface{}) fx.Option {
+	inFxType, outFxType, hasErrRet, err := constructFxInAndOut(compCtorFunc)
+	if err != nil {
+		return fx.Error(err)
+	}
+
+	// type-check the input argument to the constructor
+	constructorFunc := reflect.ValueOf(compCtorFunc)
+	if constructorFunc.Type().NumIn() > 0 && constructorFunc.Type().In(0).Kind() == reflect.Interface {
+		return fx.Error(errors.New("constructor must either take 0 arguments, or 1 argument as a requires struct"))
+	}
+	hasZeroArg := constructorFunc.Type().NumIn() == 0
+
+	// build reflect.Type of the constructor function that will be provided to `fx.Provide`
+	funcFxType := reflect.FuncOf([]reflect.Type{inFxType}, []reflect.Type{outFxType}, false)
+	if hasErrRet {
+		funcFxType = reflect.FuncOf([]reflect.Type{inFxType}, []reflect.Type{outFxType, errorInterface}, false)
+	}
+
+	// wrapper that receives fx-aware requirements, converts them into regular requirements, and calls the constructor
+	// constructor function value that will inform fx what the Components are
+	fxAwareProviderFunc := reflect.MakeFunc(funcFxType, func(args []reflect.Value) []reflect.Value {
+		// invoke the regular constructor with the correct arguments
+		ctorArgs := makeConstructorArgs(args[0], hasZeroArg)
+		plainOuts := constructorFunc.Call(ctorArgs)
+		// calling `constructFxInAndOut` earlier ensures that outs has exactly 1 element
+		res := []reflect.Value{makeFxAwareProvides(plainOuts[0], outFxType)}
+		if hasErrRet {
+			res = append(res, plainOuts[1])
+		}
+		return res
+	})
+
+	// NOTE: This will only work if ProvideComponentConstructor is called exactly once
+	return fx.Provide(fxAwareProviderFunc.Interface(), newFxLifecycleAdapter)
+}
+
+// get the input or output argument from a function type
+func getArg(typ reflect.Type, index int, isOut bool) reflect.Type {
+	if isOut && index < typ.NumOut() {
+		return typ.Out(index)
+	} else if !isOut && index < typ.NumIn() {
+		return typ.In(index)
+	}
+	return nil
+}
```

Comment:
I prefer combining them only because the two functions are so similar, and the code analyzing the arguments & return values reads much better. Found a way to combine them without using magic booleans, new function is `getWithinLimit`.